src/automata/RequestResponseRequester.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include "RequestResponseRequester.h"
+
+#include <iostream>
+
+namespace reactivesocket {
+
+void RequestResponseRequesterBase::subscribe(Subscriber<Payload>& subscriber) {
+  DCHECK(!consumingSubscriber_);
+  consumingSubscriber_.reset(&subscriber);
+  // FIXME
+  // Subscriber::onSubscribe is delivered externally, as it may attempt to
+  // synchronously deliver Subscriber::request.
+  // Q: why do we want to avoid Subscriber::request being delivered
+  // synchronously?
+}
+
+void RequestResponseRequesterBase::onNext(Payload request) {
+  switch (state_) {
+    case State::NEW: {
+      state_ = State::REQUESTED;
+      Frame_REQUEST_RESPONSE frame(
+          streamId_, FrameFlags_EMPTY, std::move(std::move(request)));
+      connection_->outputFrameOrEnqueue(frame.serializeOut());
+      break;
+    }
+    case State::REQUESTED:
+      break;
+    case State::CLOSED:
+      break;
+  }
+}
+
+void RequestResponseRequesterBase::request(size_t n) {
+  if (payload_) {
+    consumingSubscriber_.onNext(std::move(payload_.value()));
+    payload_.clear();
+  } else {
+    waitingForPayload_ = true;
+  }
+}
+
+void RequestResponseRequesterBase::cancel() {
+  switch (state_) {
+    case State::NEW:
+      state_ = State::CLOSED;
+      connection_->endStream(streamId_, StreamCompletionSignal::GRACEFUL);
+      break;
+    case State::REQUESTED: {
+      state_ = State::CLOSED;
+      connection_->outputFrameOrEnqueue(Frame_CANCEL(streamId_).serializeOut());
+      connection_->endStream(streamId_, StreamCompletionSignal::GRACEFUL);
+    } break;
+    case State::CLOSED:
+      break;
+  }
+}
+
+void RequestResponseRequesterBase::endStream(StreamCompletionSignal signal) {
+  switch (state_) {
+    case State::NEW:
+    case State::REQUESTED:
+      // Spontaneous ::endStream signal means an error.
+      DCHECK(StreamCompletionSignal::GRACEFUL != signal);
+      state_ = State::CLOSED;
+      break;
+    case State::CLOSED:
+      break;
+  }
+  // Q: don't really understand what the comment below means or if it applies
+  // in this case as well?
+  // FIXME: switch on signal and verify that callsites propagate stream errors
```

Comment:
I don't understand that one either. Feel free to keep it in and we will get to it later. Since you copy/pasted it, we will get more context once we will tackle this issue.